modules/placement-driver/src/main/java/org/apache/ignite/internal/raft/client/TopologyAwareRaftGroupService.java

```diff
@@ -0,0 +1,443 @@
+                                        return refreshAndGetLeaderWithTerm()
+                                                .thenAcceptAsync(leaderWithTerm -> {
+                                                    if (leaderWithTerm.leader() != null
+                                                            && appearedNode.name().equals(leaderWithTerm.leader().consistentId())) {
+                                                        serverEventHandler.onLeaderElected(appearedNode, leaderWithTerm.term());
+                                                    }
+                                                }, executor);
+                                    }
+
+                                    return CompletableFuture.completedFuture(null);
+                                }, executor);
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Starts an instance of topology aware RAFT client.
+     *
+     * @param groupId Replication group id.
+     * @param cluster Cluster service.
+     * @param factory Message factory.
+     * @param raftConfiguration RAFT configuration.
+     * @param configuration Group configuration.
+     * @param getLeader True to get the group's leader upon service creation.
+     * @param executor RPC executor.
+     * @param logicalTopologyService Logical topology service.
+     * @return Future to create a raft client.
+     */
+    public static CompletableFuture<RaftGroupService> start(
+            ReplicationGroupId groupId,
+            ClusterService cluster,
+            RaftMessagesFactory factory,
+            RaftConfiguration raftConfiguration,
+            PeersAndLearners configuration,
+            boolean getLeader,
+            ScheduledExecutorService executor,
+            LogicalTopologyService logicalTopologyService
+    ) {
+        return RaftGroupServiceImpl.start(groupId, cluster, factory, raftConfiguration, configuration, getLeader, executor)
+                .thenApply(raftGroupService -> new TopologyAwareRaftGroupService(cluster, factory, executor, raftConfiguration,
+                        raftGroupService, logicalTopologyService));
+    }
+
+    /**
+     * Sends a subscribe message to a specific node of the cluster.
+     *
+     * @param node Node.
+     * @param msg  Subscribe message.
+     * @return A future that will complete when the message is sent.
+     */
+    private CompletableFuture<Boolean> sendSubscribeMessage(ClusterNode node, SubscriptionLeaderChangeRequest msg) {
+        var msgSendFut = new CompletableFuture<Boolean>();
+
+        sendWithRetry(node, msg, msgSendFut);
+
+        return msgSendFut;
+    }
+
+    /**
+     * Tries to send a subscribe message until the node leaves of the cluster.
+     *
+     * @param node       Node.
+     * @param msg        Subscribe message to send.
+     * @param msgSendFut Future that will completed when the message is send or an issue prevent that.
```

Comment:
Am I right that that future resul is true when message is sent, false when the recipient node left topology and completes exceptionally in other cases? shouldn't we describe it in javadoc?

Reply:
Added description.